docs/proposals/12-oidc-webhook-authenticator.md

```diff
@@ -0,0 +1,305 @@
+1. OWA fetches the keys from `jwks_uri` and stores them in the status of `OpenIDConnect`:
+
+    ```yaml
+    apiVersion: authentication.gardener.cloud/v1alpha1
+    kind: OpenIDConnect
+    metadata:
+      name: foo
+    spec:
+      issuerURL: https://foo.bar
+      ...
+    status:
+      keys: f31deA9b... #the content of jwks_uri base64-encoded
+    ```
+
+1. OWA uses those keys, issuer, client_id and other settings to add OIDC authenticator to a list of already existing authenticators.
+
+![alt text](assets/registration.svg "Authentication with OIDC webhook")
+
+### End-user authentication via new OpenIDConnect IDP
+
+When a user presents an `id_token` obtained from a OpenID Connect the flow looks like this:
+
+1. The user authenticates in Custom IDP.
+1. `id_token` is obtained from Custom IDP.
+1. The user uses `id_token` to perform an API call to Kube APIServer.
+1. As the `id_token` is not matched by any build-in or configured authenticators in the Kube APIServer, it is send to OWA for validation.
+
+    ```json
+    {
+      "TokenReview": {
+        "kind": "TokenReview",
+        "apiVersion": "authentication.k8s.io/v1beta1",
+        "spec": {
+          "token": "ddeewfwef..."
+        }
+      }
+    }
+    ```
+
+1. OWA uses `TokenReview` to authenticate the calling API server (the Kube APIServer for delegation of authentication and authorization is different from the calling API server)
+
+    ```json
+    {
+      "TokenReview": {
+        "kind": "TokenReview",
+        "apiVersion": "authentication.k8s.io/v1beta1",
+        "spec": {
+          "token": "api-server-token..."
+        }
+      }
+    }
+    ```
+
+1. After the Authentication API server returns the identity of callee API server:
+
+    ```json
+    {
+        "apiVersion": "authentication.k8s.io/v1",
+        "kind": "TokenReview",
+        "metadata": {
+            "creationTimestamp": null
+        },
+        "spec": {
+            "token": "eyJhbGciOiJSUzI1NiIsImtpZCI6InJocEdLTXZlYjV1OE5heD..."
+        },
+        "status": {
+            "authenticated": true,
+            "user": {
+                "groups": [
+                    "system:serviceaccounts",
+                    "system:serviceaccounts:shoot--abcd",
+                    "system:authenticated"
+                ],
+                "uid": "14db103e-88bb-4fb3-8efd-ca9bec91c7bf",
+                "username": "system:serviceaccount:shoot--abcd:kube-apiserver"
+            }
+        }
+    }
+    ```
+
+    OWA makes a `SubjectAccessReview` call to the Authorization API server:
+
+    ```json
+    {
+      "apiVersion": "authorization.k8s.io/v1",
+      "kind": "SubjectAccessReview",
+      "spec": {
+        "groups": [
+          "system:serviceaccounts",
+          "system:serviceaccounts:shoot--abcd",
+          "system:authenticated"
+        ],
+        "nonResourceAttributes": {
+          "path": "/validate-token",
```

Comment:
Can you maybe add a sentence describing the reason for the `SubjectAccessReview` being to check if the Kube APIServer is actually allowed to validate tokens?